nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/PackageManagerImpl.java

```diff
@@ -480,58 +503,19 @@ public void registerSource(PackageSource source, boolean local) {
-        List<DownloadablePackage> toUpdate = new ArrayList<>();
-        List<String> toUpdateIds = new ArrayList<>();
-
-        // take all the available packages that correspond to an upgrade of an
-        // active local package and match the target platform
-        for (DownloadablePackage pkg : localPackages) {
-            // Ignore upgrade check for unused packages
-            if (!pkg.getPackageState().isInstalled()) {
-                continue;
-            }
-            for (DownloadablePackage availablePackage : availablePackages) {
-                if (availablePackage.getName().equals(pkg.getName())) {
-                    if (availablePackage.getVersion().isUpgradeFor(pkg.getVersion())) {
-                        toUpdate.add(availablePackage);
-                        toUpdateIds.add(availablePackage.getId());
-                    }
-                    break;
-                }
-            }
+        List<String> installedPackagesNames = listInstalledPackagesNames(type);
+        List<String> hotfixesNames = null;
+        if(type == null || type == PackageType.HOT_FIX){
+            // list last version of available hot-fixes too
+            hotfixesNames = listHotfixesNames(targetPlatform, CUDFHelper.defaultAllowSNAPSHOT);
+            hotfixesNames.removeAll(installedPackagesNames);
         }
+        DependencyResolution resolution = resolveDependencies(hotfixesNames, null, installedPackagesNames, targetPlatform, CUDFHelper.defaultAllowSNAPSHOT);
 
-        if (type == null || type == PackageType.HOT_FIX) {
-            // force addition of hot-fixes
-            List<DownloadablePackage> hotFixes = doMergePackages(getAllSources(), PackageType.HOT_FIX, targetPlatform);
-            for (DownloadablePackage pkg : hotFixes) {
-                // check it is not already in update list
-                if (!toUpdateIds.contains(pkg.getId())) {
-                    // check if package is not already in local
-                    boolean alreadyInLocal = false;
-                    for (DownloadablePackage lpkg : localPackages) {
-                        if (lpkg.getName().equals(pkg.getName())) {
-                            if (lpkg.getVersion().greaterOrEqualThan(pkg.getVersion())) {
-                                if (lpkg.getPackageState().isInstalled()) {
-                                    alreadyInLocal = true;
-                                }
-                            }
-                            break;
-                        }
-                    }
-                    if (!alreadyInLocal) {
-                        toUpdate.add(0, pkg);
-                    }
-                }
-            }
-        }
+        List<String> toUpdateIds = resolution.getOrderedPackageIdsToInstall();
+        List<DownloadablePackage> toUpdate = new ArrayList<>();
+        toUpdateIds.forEach(pkgId -> toUpdate.add(getPackage(pkgId)));
```

Comment:
@kevinleturc worth to add a section into https://doc.nuxeo.com/display/CORG/Coding+and+Design+Guidelines